src/openrct2/object/TerrainSurfaceObject.cpp

```diff
@@ -136,24 +138,43 @@ void TerrainSurfaceObject::ReadJson(IReadObjectContext* context, json_t& root)
     PopulateTablesFromJson(context, root);
 }
 
-uint32_t TerrainSurfaceObject::GetImageId(
-    const CoordsXY& position, int32_t length, int32_t rotation, int32_t offset, bool grid, bool underground) const
+ImageId TerrainSurfaceObject::GetImageId(
+    const CoordsXY& position, uint8_t length, uint8_t rotation, uint8_t offset, bool grid, bool underground) const
 {
-    uint32_t result = (underground ? DefaultUndergroundEntry : (grid ? DefaultGridEntry : DefaultEntry));
+    uint32_t result = DefaultEntry;
+    const std::vector<SpecialEntry>* entries = &SpecialEntries;
+    if (underground)
+    {
+        result = DefaultUndergroundEntry;
+        entries = &SpecialEntriesUnderground;
+    }
+    else if (grid)
+    {
+        result = DefaultGridEntry;
+        entries = &SpecialEntriesGrid;
+    }
+
+    TileCoordsXY tilePos(position);
+    uint8_t variation = ((tilePos.x << 1) & 0b10) | (tilePos.y & 0b01);
 
     // Look for a matching special
-    auto variation = ((position.x << 1) & 0b10) | (position.y & 0b01);
-    for (const auto& special : SpecialEntries)
+    for (const auto& special : *entries)
```

Comment:
Using std::span would become.
```suggestion
    for (const auto& special : entries)
```